Expose internal rust interface to DenseLayout

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit makes a small change to the rust code for DenseLayout that enables calling it more easily from rust. The primary obstacle was the pyfunction used PyReadonlyArray2<f64> inputs which precludes calling it with rust constructed Array2Views<f64>. This adds a new inner public function which takes the array view directly and then the pyfunction's only job is to convert the inputs and outputs to Python. The python side of the function is still building a sparse matrix and then runs reverse Cuthill–McKee to get a permutation of the densest subgraph so any rust consumers will want to keep that in mind (and maybe use sprs to do the same).

At the same time it corrects an oversight in the original implementation where the returned numpy arrays of the densest subgraph are copied instead of being returned as references. This should slightly improve performance by eliminating 3 array copies that weren't needed.

### Details and comments

mtreinish committed:
Expose internal rust interface to DenseLayout

This commit makes a small change to the rust code for DenseLayout that
enables calling it more easily from rust. The primary obstacle was the
pyfunction used PyReadonlyArray2<f64> inputs which precludes calling it
with rust constructed Array2Views<f64>. This adds a new inner public
function which takes the array view directly and then the pyfunction's
only job is to convert the inputs and outputs to Python. The python side
of the function is still building a sparse matrix and then runs reverse
Cuthill–McKee to get a permutation of the densest subgraph so any rust
consumers will want to keep that in mind (and maybe use sprs to do the
same).

At the same time it corrects an oversight in the original implementation
where the returned numpy arrays of the densest subgraph are copied
instead of being returned as references. This should slightly improve
performance by eliminating 3 array copies that weren't needed.

jakelishman reviewed (commented):
This looks fine to me.  Do either of these functions actually need to return `PyResult`? They both appear to be infallible.

Let's do this after #12121 merges, though, since there's a couple of tweaks that'll be needed.

mtreinish committed:
Merge branch 'main' into dense-layout-rust-interface

mtreinish committed:
Remove PyResult

mtreinish commented:
No, the `PyResult` wasn't needed. Removed it in https://github.com/Qiskit/qiskit/pull/12104/commits/33493a0e8858a7ccad3952caaba071fcf02e3333 and also updated it to use the bounds api.

henryzou50 reviewed (commented):
LGTM. The changes improve the Rust interface and optimize performance by reducing unnecessary array copies. Great work!

henryzou50 committed:
Merge branch 'main' into dense-layout-rust-interface
